File: python/scripts/week-02/explore-readiness-report.py
"""
Week 2 — Validate a ServiceReadinessReport from mock data.

Loads one JSON scenario and validates it against the schema.
This is your starting point. Your self-learning assignment is below.

Run: python scripts/week-02/explore-readiness-report.py
"""
import os
import sys
import json

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))

from src.schemas import ServiceReadinessReport

DATA_DIR = os.path.join(
    os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))),
    "..", "shared", "data",
)

print("=" * 70)
print("  ServiceReadinessReport — Reference Validation")
print("=" * 70)
print()

# ── Reference: healthy scenario ───────────────────────────────
path = os.path.join(DATA_DIR, "service-readiness-healthy.json")
print(f"  Loading {path}...")

with open(path) as f:
    data = json.load(f)

report = ServiceReadinessReport.model_validate(data)
print(f"  ✅ Validated — {type(report).__name__}(")
print(f"       service.name   = {report.service.name}")
print(f"       service.version= {report.service.version}")
print(f"       build.status   = {report.build.status}")
print(f"       deploy.history = {len(report.deployment.recent_deploys)} deploys")
print(f"       verdict.ready  = {report.verdict.ready}")
print(f"       verdict.conf   = {report.verdict.confidence}")
print(f"       blockers       = {report.verdict.blockers}")
print(f"       evidence       = {len(report.evidence)} chunks")
print(f"     )")

print()


# ═══════════════════════════════════════════════════════════════════
# TAKE-HOME ASSIGNMENT
# ═══════════════════════════════════════════════════════════════════
#
# PART A — Load the other two scenarios:
#   1. Extend this script to also load and validate:
#      shared/data/service-readiness-degraded.json
#      shared/data/service-readiness-unknown.json
#   2. Add the same tests to tests/test_schemas.py
#      (follow the healthy example that's already there)
#
# PART B — LLM integration:
#   Use src.schemas.generate_readiness_report() to feed
#   the mock data to the LLM and get back a typed report.
#   Example:
#
#     import json
#     from src.schemas import generate_readiness_report
#
#     with open('../shared/data/service-readiness-healthy.json') as f:
#         data = json.load(f)
#
#     report = generate_readiness_report(
#         service_name=data['service']['name'],
#         build_data=data['build'],
#         deploy_data=data['deployment'],
#         temperature=0.0,
#     )
#
#     # report is a typed ServiceReadinessReport — no parsing needed!
#     print(report.model_dump_json(indent=2))
#
#   Run this for all 3 scenarios. Compare the LLM's verdict
#   to the hand-written JSON — does the model agree? Where
#   does it differ? What would you change in the prompt?
# ═══════════════════════════════════════════════════════════════════

File: nodejs/scripts/week-02/explore-readiness-report.js
/**
 * Week 2 — Validate a ServiceReadinessReport from mock data.
 *
 * Loads one JSON scenario and validates it against the schema.
 * This is your starting point. Your self-learning assignment is below.
 *
 * Run: node scripts/week-02/explore-readiness-report.js
 */
import { readFileSync } from "fs";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";
import { ServiceReadinessReportSchema } from "../../src/schemas.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const DATA_DIR = resolve(__dirname, "..", "..", "..", "shared", "data");

console.log("=".repeat(70));
console.log("  ServiceReadinessReport — Reference Validation");
console.log("=".repeat(70));
console.log();

// ── Reference: healthy scenario ───────────────────────────────
const path = resolve(DATA_DIR, "service-readiness-healthy.json");
console.log(`  Loading ${path}...`);

const data = JSON.parse(readFileSync(path, "utf-8"));

const report = ServiceReadinessReportSchema.parse(data);
console.log(`  ✅ Validated — ${report.constructor.name}(`);
console.log(`       service.name   = ${report.service.name}`);
console.log(`       service.version= ${report.service.version}`);
console.log(`       build.status   = ${report.build.status}`);
console.log(
  `       deploy.history = ${report.deployment.recent_deploys.length} deploys`
);
console.log(`       verdict.ready  = ${report.verdict.ready}`);
console.log(`       verdict.conf   = ${report.verdict.confidence}`);
console.log(
  `       blockers       = [${report.verdict.blockers.join(", ")}]`
);
console.log(`       evidence       = ${report.evidence.length} chunks`);
console.log("     )");
console.log();

// ═══════════════════════════════════════════════════════════════════
// TAKE-HOME ASSIGNMENT
// ═══════════════════════════════════════════════════════════════════
//
// PART A — Load the other two scenarios:
//   1. Extend this script to also load and validate:
//      shared/data/service-readiness-degraded.json
//      shared/data/service-readiness-unknown.json
//   2. Tests are already in tests/test_schemas.js
//      (the degraded and unknown scenario tests are provided)
//
// PART B — LLM integration:
//   Use generateReadinessReport() from src/llm_functions.js to feed
//   the mock data to the LLM and get back a typed report.
//
//   Example:
//
//     import { generateReadinessReport } from
//       '../../src/llm_functions.js';
//
//     const data = JSON.parse(
//       readFileSync('../../shared/data/
//         service-readiness-healthy.json', 'utf-8'));
//
//     const report = await generateReadinessReport({
//       serviceName: data.service.name,
//       buildData: data.build,
//       deployData: data.deployment,
//       temperature: 0.0,
//     });
//
//     console.log(JSON.stringify(report, null, 2));
//
//   Run this for all 3 scenarios. Compare the LLM's verdict
//   to the hand-written JSON — does the model agree? Where
//   does it differ? What would you change in the system prompt?
